examples/gno.land/p/moul/collection/collection.gno

```diff
@@ -0,0 +1,314 @@
+	}()
+
+	return fn(obj), true
+}
+
+// Set adds or updates an object in the collection
+func (c *Collection) Set(obj interface{}) uint64 {
+	if obj == nil {
+		return 0
+	}
+
+	// Generate new ID
+	id := c.idGen.Next()
+	idStr := id.String()
+
+	// Check uniqueness constraints first
+	for name, idx := range c.indexes {
+		if name == IDIndex {
+			continue
+		}
+		key, ok := safeGenerateKey(idx.fn, obj)
+		if !ok {
+			return 0
+		}
+
+		// Skip empty values for sparse indexes
+		if idx.options&SparseIndex != 0 && key == "" {
+			continue
+		}
+
+		if idx.options&CaseInsensitiveIndex != 0 {
+			key = strings.ToLower(key)
+		}
+
+		// Only check uniqueness for unique indexes
+		if idx.options&UniqueIndex != 0 {
+			if existing, exists := idx.tree.Get(key); exists && existing != nil {
+				return 0 // Uniqueness constraint violated
+			}
+		}
+	}
+
+	// Store in _id index first
+	c.indexes[IDIndex].tree.Set(idStr, obj)
+
+	// Store in all other indexes
+	for name, idx := range c.indexes {
+		if name == IDIndex {
+			continue
+		}
+		key, ok := safeGenerateKey(idx.fn, obj)
+		if !ok {
+			// Rollback: remove from _id index
+			c.indexes[IDIndex].tree.Remove(idStr)
+			return 0
+		}
+
+		// Skip empty values for sparse indexes
+		if idx.options&SparseIndex != 0 && key == "" {
+			continue
+		}
+
+		if idx.options&CaseInsensitiveIndex != 0 {
+			key = strings.ToLower(key)
+		}
+
+		// For non-unique indexes, we store the ID as the value
+		idx.tree.Set(key, idStr)
+	}
+
+	return uint64(id)
+}
+
+// Get retrieves an object by index and key, returns (object, id)
+func (c *Collection) Get(indexName, key string) (interface{}, uint64) {
+	idx, exists := c.indexes[indexName]
+	if !exists {
+		return nil, 0
+	}
+
+	if indexName == IDIndex {
+		obj, exists := idx.tree.Get(key)
+		if !exists {
+			return nil, 0
+		}
+		id, err := seqid.FromString(key)
+		if err != nil {
+			return nil, 0
+		}
+		return obj, uint64(id)
+	}
+
+	// For other indexes
+	if idx.options&CaseInsensitiveIndex != 0 {
+		key = strings.ToLower(key)
+	}
+
+	idStr, exists := idx.tree.Get(key)
+	if !exists {
+		return nil, 0
+	}
+
+	// Get the actual object from _id index
+	obj, exists := c.indexes[IDIndex].tree.Get(idStr.(string))
+	if !exists {
+		return nil, 0
+	}
+
+	id, err := seqid.FromString(idStr.(string))
+	if err != nil {
+		return nil, 0
+	}
+	return obj, uint64(id)
+}
+
+// GetIndex returns the underlying tree for an index
+func (c *Collection) GetIndex(name string) avl.ITree {
+	idx, exists := c.indexes[name]
+	if !exists {
+		return nil
+	}
+	return idx.tree
+}
+
+// Delete removes an object by its ID
+func (c *Collection) Delete(id uint64) {
+	idStr := seqid.ID(id).String()
+
+	// Get the object first to clean up other indexes
+	obj, exists := c.indexes[IDIndex].tree.Get(idStr)
+	if !exists {
+		return
+	}
+
+	// Remove from all indexes
+	for name, idx := range c.indexes {
+		if name == IDIndex {
+			idx.tree.Remove(idStr)
+			continue
+		}
+		key := idx.fn(obj)
```

Comment:
In the `Set` function, when inserting data with the `CaseInsensitiveIndex` option enabled, keys are stored after being converted to lowercase.

```go
if idx.options&CaseInsensitiveIndex != 0 {
	key = strings.ToLower(key)
}
```
However, it seems that the same processing is missing when removing these keys in the `Delete` or `Update` functions.

In this line, the code attempts to directly delete the key obtained from `idx.fn(obj)` without any coversion process. I think this can lead to a unexpected behaviour which delection or updates don't work properly because the key doesn't match the actual lowercase key stored in the index.

Reply:
aa7c3aebc